File: 86. MinimumAbsoluteDifferenceInBST.java
class TreeNodeMinDiff {
    int val;
    TreeNodeMinDiff left;
    TreeNodeMinDiff right;

    TreeNodeMinDiff(int val) {
        this.val = val;
    }
}

class Solution {
    private Integer prev;
    private int minDiff;

    public int getMinimumDifference(TreeNodeMinDiff root) {
        prev = null;
        minDiff = Integer.MAX_VALUE;
        inorder(root);
        return minDiff;
    }

    private void inorder(TreeNodeMinDiff node) {
        if (node == null) {
            return;
        }

        inorder(node.left);

        if (prev != null) {
            minDiff = Math.min(minDiff, node.val - prev);
        }
        prev = node.val;

        inorder(node.right);
    }
}

public class MinimumAbsoluteDifferenceInBST {
    public static void main(String[] args) {
        Solution solution = new Solution();

        TreeNodeMinDiff root1 = new TreeNodeMinDiff(4);
        root1.left = new TreeNodeMinDiff(2);
        root1.right = new TreeNodeMinDiff(6);
        root1.left.left = new TreeNodeMinDiff(1);
        root1.left.right = new TreeNodeMinDiff(3);

        TreeNodeMinDiff root2 = new TreeNodeMinDiff(1);
        root2.right = new TreeNodeMinDiff(3);
        root2.right.left = new TreeNodeMinDiff(2);

        System.out.println(solution.getMinimumDifference(root1));
        System.out.println(solution.getMinimumDifference(root2));
    }
}

File: 86. MinimumAbsoluteDifferenceInBST.c
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

struct TreeNode *newNode(int val) {
    struct TreeNode *node = malloc(sizeof(*node));
    node->val = val;
    node->left = node->right = NULL;
    return node;
}

void inorder(struct TreeNode *root, int *prev, int *best) {
    if (!root) return;
    inorder(root->left, prev, best);
    if (*prev != INT_MIN && root->val - *prev < *best) *best = root->val - *prev;
    *prev = root->val;
    inorder(root->right, prev, best);
}

int getMinimumDifference(struct TreeNode *root) {
    int prev = INT_MIN, best = INT_MAX;
    inorder(root, &prev, &best);
    return best;
}

void freeTree(struct TreeNode *root) {
    if (!root) return;
    freeTree(root->left);
    freeTree(root->right);
    free(root);
}

int main(void) {
    struct TreeNode *root1 = newNode(4);
    root1->left  = newNode(2);
    root1->right = newNode(6);
    root1->left->left  = newNode(1);
    root1->left->right = newNode(3);

    struct TreeNode *root2 = newNode(1);
    root2->right = newNode(3);
    root2->right->left = newNode(2);

    printf("%d\n", getMinimumDifference(root1));
    printf("%d\n", getMinimumDifference(root2));

    freeTree(root1);
    freeTree(root2);
    return 0;
}
